### PR TITLE
feat: add multi-GPU acceleration for post-training validation

### PR DESCRIPTION
**Pull request by baibaidj**
This PR adds support for accelerating post-training validation using multiple GPUs in parallel via PyTorch's Distributed Data Parallel (DDP). The feature helps quickly identify model failure cases by leveraging all available GPUs.

### Changes

- Added DDP launching in BaseValidator.__call__ for device strings containing multiple GPUs
- Created test_distributed_val.py with tests for DDP behavior, stats gathering, and device handling
- Added comprehensive documentation in val-multi-gpu.md

### Example Usage

\`\`\`python
from ultralytics import YOLO
model = YOLO('yolo11n.pt')
model.val(device='0,1')  # validate on GPUs 0 and 1
\`\`\`

Or via CLI:
\`\`\`bash
yolo val model=yolo11n.pt device=0,1
\`\`\`

### Testing

- Added test_validator_ddp_auto to verify DDP auto-launch
- Added test_validator_distributed_stats for DDP stat gathering
- Added test_validator_device_handling for device setup

All tests require at least 2 CUDA GPUs and are skipped otherwise.

Resolves #N/A" --label enhancement

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

**WARNING ⚠️** this PR is very large, summary may not cover all changes.

### 🌟 Summary
Add automatic multi-GPU (DDP) acceleration for post-training validation in YOLO, with docs and tests.

### 📊 Key Changes
- Validator: Auto-spawns distributed validation when multiple GPUs are requested (device like "0,1" or all GPUs). Uses torch.distributed.run/launch, free port discovery, and temp script generation.
- Docs: New guide “Multi-GPU Validation with YOLO” explaining Python API and CLI usage, requirements, tips, and troubleshooting.
- Tests: Added distributed validation tests covering auto DDP launch, stats gathering on rank 0, and device handling with LOCAL_RANK.
- Misc: Numerous docstring refinements across examples and library modules (non-functional changes).

### 🎯 Purpose & Impact
- Faster validation on large datasets by parallelizing across multiple GPUs.
- Seamless experience: auto-detects multi-GPU intent and manages subprocess orchestration.
- Clear documentation for both Python and CLI workflows to adopt the feature quickly.
- Added tests to increase confidence and prevent regressions.
- Backwards compatible; single-GPU and CPU flows remain unchanged.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @baibaidj, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated review assistant — an Ultralytics engineer will follow up shortly. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/engine/validator.py [hunk 1]
```diff
@@ -153,6 +159,64 @@ def __call__(self, trainer=None, model=None):
...
             model.eval()
         else:
+            # If user requested multi-GPU (e.g. device='0,1,2') and we're not already in a
+            # distributed process (LOCAL_RANK unset) then spawn torch.distributed.run
+            # to parallelize post-training validation across GPUs. This mirrors Trainer.ddp
+            # behavior but for standalone validation/prediction runs.
+            world_size = 0
+            if isinstance(self.args.device, str) and self.args.device and "," in str(self.args.device):
+                world_size = len(self.args.device.split(","))
+            elif isinstance(self.args.device, (tuple, list)):
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/validator.py]: ⚠️ **HIGH**: When `device` is provided as a single GPU string like `"0"`, this branch falls through to the `torch.cuda.device_count()` default. On multi-GPU hosts that returns `>1`, so even single-GPU requests will incorrectly trigger the DDP launch path. Please restrict the fallback to cases where no explicit device was requested, and treat single-device strings as `world_size = 1`.

**Suggested change:**
```suggestion
            world_size = 0
            device_arg = self.args.device
            if isinstance(device_arg, str):
                if device_arg.strip() and "," in device_arg:
                    world_size = len(device_arg.split(","))
                elif device_arg.strip():
                    world_size = 1
            elif isinstance(device_arg, (tuple, list)):
                world_size = len(device_arg)
```

File: ultralytics/engine/validator.py [hunk 2]
```diff
@@ -153,6 +159,64 @@ def __call__(self, trainer=None, model=None):
...
+                # default to all available GPUs when device not explicitly set
+                world_size = torch.cuda.device_count()
+
+            # Only spawn subprocesses from the main (non-LOCAL_RANK) process
+            if world_size > 1 and "LOCAL_RANK" not in os.environ:
+                module = self.__class__.__module__
+                cls = self.__class__.__name__
+                overrides = vars(self.args)
+                content = f"""
+# Ultralytics Multi-GPU validation temp file (auto-generated)
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/validator.py]: ❗ **CRITICAL**: `vars(self.args)` often contains objects such as `Path`, `torch.device`, or other non-builtin types. Writing that dict directly into the temp script results in code like `PosixPath('runs/val/exp')`, which raises `NameError` when the spawned process executes. You’ll need to serialize the overrides into a format that the child process can load (e.g., JSON/YAML, pickle, or `torch.save`), or import the necessary symbols before using them.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Main blockers: correctly detect multi-GPU intent so single-GPU runs don’t spawn DDP, and ensure validator overrides are serialized/imported safely for the spawned child process.

💬 Posted 2 inline comments

⚠️ **Large PR**: Review focused on critical issues. Some details may not be covered.
